### PR TITLE
Kernel: Allow configuring a Jail to not impose PID isolation restriction 

### PR DESCRIPTION
**Pull request by supercomputer7**
This is quite useful for userspace applications that can't cope with the
restriction, but it's still useful to impose other non-configurable
restrictions by using jails.

Relies on #16980

### CODE REVIEW

**Comment by supercomputer7** [comment]
I am still not sure how we can ensure that userspace applications will be started inside a jail automatically, maybe some commandline embedding is required, maybe embedding special tags in the executables so the DynamicLoader will also attach itself to a PID-isolation-less jail.

**Comment by supercomputer7** [comment]
This is not stale.

**Comment by supercomputer7** [comment]
> 6:47 PM]supercomputer7: there was another idea with jail leader, which is essentially a process that can communicate with the jailed processes, but it is not jailed
> [6:47 PM]supercomputer7: so WindowServer could be a jail leader of many jails with GUI apps in them
> [6:49 PM]supercomputer7: however, I am not sure if that's enough for GUI apps (or most of them) to run in a jail. adding more than one process to this role feels just wrong so if I find that we need more programs in multiple jails (which essentially breaks pid isolation) I will just go for PID isolation-less feature

https://discord.com/channels/830522505605283862/830525093905170492/1084155649417547877

Just to keep track of ideas around this.

**Comment by supercomputer7** [comment]
This is not stale.

File: Kernel/API/Jail.h [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+#include <AK/EnumBits.h>
+#include <AK/Types.h>
+
+namespace Kernel {
+
+enum class JailFlags : u32 {
+    None = 0,
+    NoPIDIsolation = 1 << 0,
+};
```
Comment (GMTA) [inline Kernel/API/Jail.h:17]: In terms of design; conceptually jails could have a number of isolation features in the future that arguably can all be turned on or off. Instead of using the negative form here for flags ("no PID isolation") and specific APIs (`create_without_process_list`), how about using the positive approach by passing in `JailIsolationFlags` such as `::PIDIsolation` and only performing the isolation checks if a certain check is set? You could then default to a special flag by default that sets all isolation features.

This feels more "natural" to me than individually disabling features.

Comment (supercomputer7) [reply Kernel/API/Jail.h:17]: While I agree it feels more natural to some users, I never designed the jail code so you would disable many features at once. I still need to look how it would look with this because this was not the original design in my mind (but it could be the "new" design if it's neat enough!).

File: Userland/Libraries/LibCore/System.cpp [hunk 2]
```diff
@@ -1172,9 +1172,19 @@ ErrorOr<void> join_jail(u64 jail_index)
     HANDLE_SYSCALL_RETURN_VALUE("jail_attach", rc, {});
 }
 
-ErrorOr<u64> create_jail(StringView jail_name)
+ErrorOr<u64> create_jail_with_default_isolation_options(StringView jail_name)
 {
-    Syscall::SC_jail_create_params params { 0, { jail_name.characters_without_null_termination(), jail_name.length() } };
+    Syscall::SC_jail_create_params params { 0, { jail_name.characters_without_null_termination(), jail_name.length() }, 1 << 0 };
```
Comment (GMTA) [inline Userland/Libraries/LibCore/System.cpp:1177]: In terms of using the jails API from userspace, it would be nice to have the flags available here as well instead of constants like `1 << 0`.

File: Userland/Utilities/jail-create.cpp [hunk 3]
```diff
@@ -13,15 +13,21 @@ ErrorOr<int> serenity_main(Main::Arguments arguments)
...
-        return 0;
-    }
+    if (new_jail_name.is_null() || new_jail_name.is_empty())
+        return Error::from_string_view("Can't create a jail with empty name."sv);
 
-    return Error::from_string_view("Can't create a jail with empty name."sv);
+    if (custom_isolation_options)
+        TRY(Core::System::create_jail_with_custom_isolation_options(new_jail_name, pid_isolation));
+    else
+        TRY(Core::System::create_jail_with_default_isolation_options(new_jail_name));
```
Comment (GMTA) [inline Userland/Utilities/jail-create.cpp:31]: Instead of splitting all these methods into two versions ("default" and "custom options"), why not have a single method that takes the isolation flags and let userland (in this case, `jail-create`) determine what flags should be used?

File: Kernel/Jail.h [hunk 4]
```diff
@@ -52,7 +56,9 @@ class Jail : public AtomicRefCounted<Jail> {
     using List = IntrusiveListRelaxedConst<&Jail::m_list_node>;
 
 private:
-    NonnullRefPtr<ProcessList> const m_process_list;
+    // NOTE: Although we access this RefPtr during runtime multiple times, we never set
+    // this pointer after construction so it's safe to leave this without any kind of locking.
```
Comment (GMTA) [inline Kernel/Jail.h:60]: This comment seems unrelated and maybe a bit too much since the `RefPtr` is already `const`?

File: Kernel/Jail.cpp [hunk 5]
```diff
@@ -19,16 +19,19 @@ static JailIndex generate_jail_id()
...
 }
 
-NonnullRefPtr<ProcessList> Jail::process_list()
+RefPtr<ProcessList> Jail::process_list()
 {
     return m_process_list;
 }
 
-ErrorOr<NonnullRefPtr<Jail>> Jail::create(NonnullOwnPtr<KString> name)
+ErrorOr<NonnullRefPtr<Jail>> Jail::create(NonnullOwnPtr<KString> name, EnforcePIDIsolation enforce_pid_isolation)
```
Comment (GMTA) [inline Kernel/Jail.cpp:27]: Can't we accept the flags here as well? This tends to grow in number of arguments as soon as we think of other isolation options (networking, filesystem, etc.). I can imagine the invoker would translate the `::Default` flag to the individual default isolation option flags and passing those to this method.

File: Kernel/Syscalls/jail.cpp [hunk 6]
```diff
@@ -14,6 +14,18 @@ namespace Kernel {
 
 constexpr size_t jail_name_max_size = 50;
 
+static ErrorOr<NonnullRefPtr<Jail>> create_jail_with_default_isolation_options(NonnullOwnPtr<KString> jail_name)
+{
+    return TRY(Jail::create(move(jail_name), Jail::EnforcePIDIsolation::Yes));
+}
+
+static ErrorOr<NonnullRefPtr<Jail>> create_jail_with_custom_isolation_options(NonnullOwnPtr<KString> jail_name, unsigned flags)
```
Comment (GMTA) [inline Kernel/Syscalls/jail.cpp:22]:
```cpp
static ErrorOr<NonnullRefPtr<Jail>> create_jail_with_custom_isolation_options(NonnullOwnPtr<KString> jail_name, u32 flags)
```
This reflects the type in use for `JailIsolationFlags`. Should alleviate the need for all the `static_cast`s I think?

File: Kernel/Syscalls/jail.cpp [hunk 7]
```diff
@@ -24,14 +36,25 @@ ErrorOr<FlatPtr> Process::sys$jail_create(Userspace<Syscall::SC_jail_create_para
     if (jail_name->length() > jail_name_max_size)
         return ENAMETOOLONG;
 
+    // NOTE: Either the default isolation flag is set or individual flags are set
+    if (params.flags & static_cast<unsigned>(JailIsolationFlags::Default)
+        && ((params.flags & ~static_cast<unsigned>(JailIsolationFlags::Default)) != 0))
```
Comment (GMTA) [inline Kernel/Syscalls/jail.cpp:41]: This means `::None` is an invalid flag - either we should support that flag, or we should not have it in the first place?

Comment (supercomputer7) [reply Kernel/Syscalls/jail.cpp:41]: Ah, I meant that if you don't set anything then that's actually OK, but if you set the default flag with other flags this is invalid. I will clarify this.

File: Userland/Utilities/jail-create.cpp (hunk 3 above)
Comment (supercomputer7) [reply Userland/Utilities/jail-create.cpp:31]: But then there's no meaning to the default flag, because having a function:
```
ErrorOr<u64> create_jail(StringView jail_name, bool default_isolation_options, bool pid_isolation)
```
does not make sense. I now just think about dropping the whole concept of "default flag", because it makes things too much complicated.

**Review by GMTA — APPROVED** [review]
This simplified it a lot, I like it!
